modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryReducer.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.query;
+
+import java.util.Collection;
+import java.util.UUID;
+import org.apache.ignite.IgniteCheckedException;
+
+/**
+ * This class is responsible for reducing results of cache query. Query results are delivered with function
+ * {@link #onPage(UUID, Collection, boolean)}. Note that this reducer deeply interacts with corresponding query future
+ * {@link GridCacheQueryFutureAdapter}, so they used the same lock object. It guards reducer pages operations
+ * and the future status. Custom reduce logic is applied within {@link #next()} and {@link #hasNext()}.
```

Comment:
```suggestion
 * This class is responsible for reducing results of cache query. 
 *
 * Query results are delivered via callback {@link #onPage(UUID, Collection, boolean)}.
 * @see GridCacheQueryFutureAdapter.
```

Reply:
Actually, interface don't imply any coupling with the  query future.
If you think that this worth to be mentioned, then abstract class or concrete implementation would be a better place.

Reply:
Actually, shared lock now is part of PageStream only. So we can remove this mention.